src/kbmod/result_list.py

```diff
@@ -359,16 +382,50 @@ def append_result_dict(self, res_dict):
             for i in range(num_results):
                 self.results[i].all_stamps = res_dict["all_stamps"][i]
 
-    def filter_results(self, indices_to_keep):
+    def filter_results(self, indices_to_keep, label=None):
         """Filter the rows in the ResultList to only include those indices
         in the list indices_to_keep.
 
         Parameters
         ----------
         indices_to_keep : list
             The indices of the rows to keep.
+        label : string
+            The label of the filtering stage to use. Only used if
+            we keep filtered trajectories.
+
+        Returns
+        -------
+        self : ResultList
+            Returns a reference to itself to allow chaining.
         """
-        self.results = [self.results[i] for i in indices_to_keep]
+        if not self.track_filtered:
+            self.results = [self.results[i] for i in indices_to_keep]
+        else:
+            keep_set = set(indices_to_keep)
+
+            # Divide the current results into a set of rows to keep
+            # and a set to filter.
+            keep_rows = []
+            filter_rows = []
+            for i, row in enumerate(self.results):
```

Comment:
I don't know what `len(self.results)` typically looks like, but python `list.append` can get pretty slow for a lot of elements. I guess this path isn't running in high-performance runs, but some kinda list comprehension could speed up this stage.

Reply:
Changed